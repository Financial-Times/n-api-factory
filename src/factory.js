import fetch from 'node-fetch';
import querystring from 'querystring';
import { parseFetchError } from '@financial-times/n-error';
import { trimObject } from '@financial-times/n-utils';

export const defaultHeaders = ({
	API_KEY,
	transactionId,
	systemId = process.env.SYSTEM_CODE,
	user = 'customer',
}) =>
	trimObject({
		'x-api-key': API_KEY,
		'FT-Transaction-Id': transactionId,
		'Content-Type': 'application/json',
		'X-Origin-System-Id': `https://cmdb.ft.com/systems/${systemId}`,
		'X-Origin-User': user,
	});

export const fetchWithDefaultErrorHandling = async (url, options) => {
	try {
		const response = await fetch(url, options);
		const contentLength = parseInt(response.headers.get('content-length'), 10);

		if (!response.ok) {
			throw response;
		}
		// in case of 'NO CONTENT'
		if (response.status === 204 || contentLength === 0) {
			return undefined;
		}
		return await response.json();
	} catch (e) {
		const parsed = await parseFetchError(e);
		throw parsed;
	}
};

export const setupMethod = method => ({ API_HOST, API_KEY }) => ({
	endpoint,
	query,
	body,
	meta,
}) => {
	const { transactionId } = meta;
	const options = trimObject({
		method,
		headers: defaultHeaders({ API_KEY, transactionId }),
		body: body ? JSON.stringify(body) : undefined,
	});
	// query fields can be nullable
	const stringifiedQuery = query
		? querystring.stringify(trimObject(query))
		: undefined;
	const url = stringifiedQuery
		? `${API_HOST}${endpoint}?${stringifiedQuery}`
		: `${API_HOST}${endpoint}`;
	return fetchWithDefaultErrorHandling(url, options);
};

export const setupService = ({ API_HOST, API_KEY }) => ({
	get: setupMethod('GET')({ API_HOST, API_KEY }),
	post: setupMethod('POST')({ API_HOST, API_KEY }),
	delete: setupMethod('DELETE')({ API_HOST, API_KEY }),
});

export default setupService;
