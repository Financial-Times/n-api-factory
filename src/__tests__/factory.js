import nock from 'nock';
import fetch from 'node-fetch';

import { fetchWithDefaultErrorHandling, setupService } from '../factory';

const config = {
	API_HOST: 'http://mock.com',
	API_KEY: 'mock.key',
};

process.env.SYSTEM_CODE = 'newspaper-mma';

describe('fetchWithDefaultErrorHandling', () => {
	afterEach(() => {
		nock.cleanAll();
	});

	it('return the parsed data correctly if response.ok', async () => {
		const res = { foo: 'bar' };
		nock(config.API_HOST)
			.get('/')
			.reply(200, res);
		const data = await fetchWithDefaultErrorHandling(`${config.API_HOST}`);
		expect(data).toEqual(res);
	});

	it('return undefined if response.status 204', async () => {
		nock(config.API_HOST)
			.get('/')
			.reply(204);

		const data = await fetchWithDefaultErrorHandling(`${config.API_HOST}`);
		expect(data).toBe(undefined);
	});

	it('return undefined if the response.body is emtpy', async () => {
		nock(config.API_HOST)
			.get('/')
			.reply(200, '', { 'Content-Length': '0' });

		const data = await fetchWithDefaultErrorHandling(`${config.API_HOST}`);
		expect(data).toBe(undefined);
	});

	it('return formatted error correctly if response is error', async () => {
		nock(config.API_HOST)
			.get('/')
			.reply(404, { message: 'some message' });
		try {
			await fetchWithDefaultErrorHandling(`${config.API_HOST}`);
			throw new Error('it should throw an exception');
		} catch (e) {
			expect(e).toEqual({
				category: 'FETCH_RESPONSE_ERROR',
				contentType: 'application/json',
				status: 404,
				message: {
					message: 'some message',
				},
			});
		}
	});
});

describe('setupService', () => {
	beforeAll(() => {
		fetch.default = jest.fn();
	});

	afterEach(() => {
		jest.resetAllMocks();
	});

	it('set up get method correctly', () => {
		const service = setupService(config);
		const transactionId = 'xxxx';
		service.get({
			endpoint: '/test',
			meta: { transactionId },
		});
		expect(fetch.default.mock.calls).toMatchSnapshot();
	});

	it('set up post methods correctly', () => {
		const service = setupService(config);
		const transactionId = 'xxxx';
		service.post({
			endpoint: '/test',
			body: { test: 'test' },
			meta: { transactionId },
		});
		expect(fetch.default.mock.calls).toMatchSnapshot();
	});

	it('set up delete methods correctly', () => {
		const service = setupService(config);
		const transactionId = 'xxxx';
		service.delete({ endpoint: '/test/xxxx', meta: { transactionId } });
		expect(fetch.default.mock.calls).toMatchSnapshot();
	});
});
