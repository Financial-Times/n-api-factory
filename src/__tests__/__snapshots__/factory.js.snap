// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setupService set up delete methods correctly 1`] = `Array []`;

exports[`setupService set up get method correctly 1`] = `Array []`;

exports[`setupService set up post methods correctly 1`] = `Array []`;
